gensim/models/nmf.py

```diff
@@ -422,33 +503,37 @@ def get_document_topics(self, bow, minimum_probability=None,
             kwargs = dict(minimum_probability=minimum_probability)
             return self._apply(corpus, **kwargs)
 
-        v = matutils.corpus2csc([bow], len(self.id2word)).tocsr()
-        h, _ = self._solveproj(v, self._W, v_max=np.inf)
+        v = matutils.corpus2csc([bow], self.num_tokens)
+        h = self._solveproj(v, self._W, v_max=np.inf)
 
         if normalize is None:
             normalize = self.normalize
         if normalize:
-            h.data /= h.sum()
+            h /= h.sum()
 
         return [
-            (idx, proba.toarray()[0, 0])
+            (idx, proba)
             for idx, proba in enumerate(h[:, 0])
-            if not minimum_probability or proba.toarray()[0, 0] > minimum_probability
+            if not minimum_probability or proba > minimum_probability
         ]
 
     def _setup(self, corpus):
         """Infer info from the first document and initialize matrices.
 
         Parameters
         ----------
-        corpus : iterable of list(int, float)
-            Training corpus.
+        corpus : iterable of list of (int, float), optional
+            Training corpus. Contains list of counts of words for every document.
 
         """
-        self._h, self._r = None, None
-        first_doc_it = itertools.tee(corpus, 1)
-        first_doc = next(first_doc_it[0])
-        first_doc = matutils.corpus2csc([first_doc], len(self.id2word))
+        self._h = None
+
+        if isinstance(corpus, scipy.sparse.csc.csc_matrix):
+            first_doc = corpus.getcol(0)
+        else:
+            first_doc_it = itertools.tee(corpus, 1)
+            first_doc = next(first_doc_it[0])
```

Comment:
I think this will still advance the original generator (lose the first document).

Isn't there a mismatch between expected corpus docs and actually seen corpus docs during iteration, after `_setup`?

Reply:
No, that's something that I checked, `itertools.tee` copies generator and leaves the original one intact.

Reply:
```python
corpus = (i for i in range(10))
first_doc_it = itertools.tee(corpus, 1)
first_doc = next(first_doc_it[0])
print("first doc", first_doc)
for doc_no, doc in enumerate(corpus):
    print(f"doc #{doc_no}: doc")

outputs:

first doc 0
doc #0: 1
doc #1: 2
doc #2: 3
doc #3: 4
doc #4: 5
doc #5: 6
doc #6: 7
doc #7: 8
doc #8: 9
```

Reply:
Does NMF accept generators? Or does it require a "restartable" sequence in `corpus`?

Reply:
Can you add those corner cases into unit tests, too? To prevent regressions. (generator input, empty docs, corpus shape not divisible by chunksize, etc)

Reply:
NMF loses the first document of a non-restartable generator for now, though non-restartable sequence should work fine with `passes=1`.

I'll fix that and add unit tests.